### PR TITLE
Add deposit return voucher states as part of the checkout process (Apps-2040)

### PR DESCRIPTION
**Pull request by Fabtron**
### What's new?

We can now track the state of deposit return vouchers in the checkout process. If a redemption failes, due to whatever reason, we receive a matching state in a new property `depositReturnVouchers` which holds the state for each drv. The state is handled and the user is informed via a Dialog that the reffering deposit return voucher will be removed to proceed with the checkout, e.g. to try again without it.

APPS-2040

### How to test?
Integrate this branch into an app using drv's and scan the product linked in the issue. Add an additional item to get to a positive amount and checkout. -> The dialog should pop up after the checkout failed.

### Definition of Done

- [x] Issue is linked
- [x] All requirements of the issue are fulfilled
- [x] Changelog is updated
- [x] Documentation is updated
- [x] [Self-Review](https://www.nerdwallet.com/blog/engineering/why-you-should-be-doing-self-reviews/)
- [ ] Review with the Product Owner _(Release-Variante o. Minified Build)_

#### App Tests
- [ ] Minified build has been tested _(aka. Release Build)_
- [ ] Environments have been taken care of _(Production/Staging)_
- [x] Supported languages have been tested
- [ ] Light-/Dark-Mode has been tested
- [ ] Edge-Cases have been tested
- [ ] Android API Levels have been taken care of _(minSdk?)_

#### Testing
- [ ] Tests have been written _(aka Unit-Tests, Integration-Tests, ...)_

### CODE REVIEW

File: core/src/main/java/io/snabble/sdk/checkout/CheckoutApi.kt [hunk 1]
```diff
@@ -334,11 +334,37 @@ data class Fulfillment(
...
+    @SerializedName("pending")
+    PENDING,
+
+    @SerializedName("redeemed")
+    REDEEMED,
+
+    @SerializedName("redeemingFailed")
+    REDEEMING_FAILED,
+
+    @SerializedName("rolledback")
```
Comment (cmaier) [inline core/src/main/java/io/snabble/sdk/checkout/CheckoutApi.kt:354]: Is the is correct that it's all lower case and not camel case?

File: core/src/main/java/io/snabble/sdk/checkout/CheckoutApi.kt [hunk 2]
```diff
@@ -334,11 +334,37 @@ data class Fulfillment(
...
+    ROLLBACK_FAILED
+}
+
 data class CheckoutProcessResponse(
     val links: Map<String, Href>? = null,
     val checks: List<Check> = emptyList(),
     @SerializedName("orderID")
     val orderId: String? = null,
+    @SerializedName("depositReturnVouchers")
+    val depositReturnVouchers: List<DepositReturnVoucher>? = null,
```
Comment (cmaier) [inline core/src/main/java/io/snabble/sdk/checkout/CheckoutApi.kt:367]: The annotation is obsolete in this case, but I don't mind, we can keep it!

File: ui/src/main/java/io/snabble/sdk/ui/cart/CheckoutBar.kt [hunk 3]
```diff
@@ -469,12 +475,42 @@ open class CheckoutBar @JvmOverloads constructor(
...
+    private fun handleFailedDepositReturns() {
+        val failedDepositReturnVouchers =
+            project.checkout.checkoutProcess?.depositReturnVouchers
+                ?.filter { it.state == DepositReturnVoucherState.REDEEMING_FAILED }
+                ?: return
+
+        val message = failedDepositReturnVouchers
+            .joinToString(separator = "/n") {
+                cart.getByItemId(it.refersTo)?.displayName.toString()
+            }
```
Comment (cmaier) [inline ui/src/main/java/io/snabble/sdk/ui/cart/CheckoutBar.kt:494]: This could lead to "null" being appended. Maybe this `cart.getByItemId(it.refersTo)?.displayName ?: ""` would be more appropriate or even this:

```kotlin
.mapNotNull { cart.getByItemId(it.refersTo)?.displayName }
.jointToString(/* ... */)
```

**Review by cmaier — CHANGES_REQUESTED** [review]
Works fine. Just minor findings.

File: core/src/main/java/io/snabble/sdk/checkout/CheckoutApi.kt (hunk 1 above)
Comment (Fabtron) [reply core/src/main/java/io/snabble/sdk/checkout/CheckoutApi.kt:354]: yes it is correct

File: ui/src/main/java/io/snabble/sdk/ui/cart/CheckoutBar.kt (hunk 3 above)
Comment (Fabtron) [reply ui/src/main/java/io/snabble/sdk/ui/cart/CheckoutBar.kt:494]: I like the second approach. Goood point changed it 👍

File: core/src/main/java/io/snabble/sdk/checkout/CheckoutApi.kt (hunk 2 above)
Comment (Fabtron) [reply core/src/main/java/io/snabble/sdk/checkout/CheckoutApi.kt:367]: 👍

**Review by cmaier — APPROVED** [review]
